BungeeCord-Patches/0054-Speed-up-packet-construction.patch

```diff
@@ -0,0 +1,347 @@
+                     EncryptionResponse.class,
++                    EncryptionResponse::new, // Waterfall - speed up packet construction
+                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
+             );
+             TO_SERVER.registerPacket(
+                     LoginPayloadResponse.class,
++                    LoginPayloadResponse::new, // Waterfall - speed up packet construction
+                     map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
+             );
+         }
+@@ -368,7 +405,7 @@ public enum Protocol
+ 
+         private final int protocolVersion;
+         private final TObjectIntMap<Class<? extends DefinedPacket>> packetMap = new TObjectIntHashMap<>( MAX_PACKET_ID );
+-        private final Constructor<? extends DefinedPacket>[] packetConstructors = new Constructor[ MAX_PACKET_ID ];
++        private final java.util.function.Supplier<? extends DefinedPacket>[] packetConstructors = new java.util.function.Supplier[ MAX_PACKET_ID ]; // Waterfall - speed up packet construction
+     }
+ 
+     @Data
+@@ -436,21 +473,24 @@ public enum Protocol
+                 throw new BadPacketException( "Packet with id " + id + " outside of range " );
+             }
+ 
+-            Constructor<? extends DefinedPacket> constructor = protocolData.packetConstructors[id];
++            java.util.function.Supplier<? extends DefinedPacket> constructor = protocolData.packetConstructors[id]; // Waterfall - speed up packet construction
+             try
+             {
+-                return ( constructor == null ) ? null : constructor.newInstance();
+-            } catch ( ReflectiveOperationException ex )
++                return ( constructor == null ) ? null : constructor.get(); // Waterfall - speed up packet construction
++            } catch ( Exception ex ) // Waterfall - speed up packet construction
+             {
+                 throw new BadPacketException( "Could not construct packet with id " + id, ex );
+             }
+         }
+ 
+-        private void registerPacket(Class<? extends DefinedPacket> packetClass, ProtocolMapping... mappings)
++        private <P extends DefinedPacket> void registerPacket(Class<P> packetClass, java.util.function.Supplier<P> constructor, ProtocolMapping... mappings) // Waterfall - speed up packet construction
+         {
++            // Waterfall start - speed up packet construction
++            /*
+             try
+             {
+                 Constructor<? extends DefinedPacket> constructor = packetClass.getDeclaredConstructor();
++                */ // Waterfall end
+ 
+                 int mappingIndex = 0;
+                 ProtocolMapping mapping = mappings[mappingIndex];
+@@ -479,11 +519,32 @@ public enum Protocol
+                     data.packetMap.put( packetClass, mapping.packetID );
+                     data.packetConstructors[mapping.packetID] = constructor;
+                 }
++            // Waterfall start - speed up packet construction
```

Comment:
Removed the try/catch, this woulda incorrectly wrapped exceptions here

Reply:
Whoops, overlooked that detail :eyes: but thanks